langchain/src/agents/executor.ts

```diff
@@ -1,25 +1,50 @@
 import { BaseChain, ChainInputs } from "../chains/base.js";
-import { BaseMultiActionAgent, BaseSingleActionAgent } from "./agent.js";
+import {
+  BaseMultiActionAgent,
+  BaseSingleActionAgent,
+  RunnableAgent,
+} from "./agent.js";
 import { StoppingMethod } from "./types.js";
 import { SerializedLLMChain } from "../chains/serde.js";
 import {
   AgentAction,
   AgentFinish,
   AgentStep,
+  BaseMessage,
   ChainValues,
 } from "../schema/index.js";
 import { CallbackManagerForChainRun } from "../callbacks/manager.js";
 import { OutputParserException } from "../schema/output_parser.js";
-import { Tool, ToolInputParsingException } from "../tools/base.js";
+import {
+  StructuredTool,
+  Tool,
+  ToolInputParsingException,
+} from "../tools/base.js";
+import { Runnable } from "../schema/runnable/base.js";
+
+type ExtractToolType<T> = T extends { ToolType: infer Tool }
+  ? Tool
+  : StructuredTool;
 
 /**
  * Interface defining the structure of input data for creating an
  * AgentExecutor. It extends ChainInputs and includes additional
  * properties specific to agent execution.
  */
-export interface AgentExecutorInput extends ChainInputs {
-  agent: BaseSingleActionAgent | BaseMultiActionAgent;
-  tools: this["agent"]["ToolType"][];
+export interface AgentExecutorInput<
+  RunInput extends ChainValues & {
+    agent_scratchpad?: string | BaseMessage[];
+    stop?: string[];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  } = any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  RunOutput extends AgentAction | AgentFinish = any
+> extends ChainInputs {
+  agent:
+    | BaseSingleActionAgent
+    | BaseMultiActionAgent
+    | Runnable<RunInput, RunOutput>;
+  tools: ExtractToolType<this["agent"]>[];
```

Comment:
Do we actually need this? Can't we just declare this on `RunnableAgent`?

Reply:
The issue is we're not taking in a `RunnableAgent`, instead we're taking in a runnable or base agent, and reassigning the runnable to `RunnableAgent`
If we want to remove `ExtractToolType` we'd need to add `ToolType` to the base `Runnable` class

Reply:
Oh right. Never mind